backend/Application/Aggregates/Contract/Entities/ContractSnapshot.cs

```diff
@@ -0,0 +1,198 @@
+        )
+    {
+        foreach (var contractAddress in addressesUpdated)
+        {
+            var address = new ContractAddress(contractAddress.ContractAddressIndex, contractAddress.ContractAddressSubIndex);
+            var contractEvents = contractEventsAdded
+                .Where(ce => ce.ContractAddressIndex == contractAddress.ContractAddressIndex && ce.ContractAddressSubIndex == contractAddress.ContractAddressSubIndex)
+                .OrderBy(ce => ce.BlockHeight)
+                .ThenBy(ce => ce.TransactionIndex)
+                .ThenBy(ce => ce.EventIndex)
+                .ToList();
+
+            var latestContractSnapshot = await repository.GetReadonlyLatestContractSnapshot(address);
+            var blockHeight = contractEvents.Last().BlockHeight;
+            var moduleReferenceEvent = GetLatestModuleReferenceContractLinkEvent(address, moduleReadonlyRepository);
+            var amount = GetAmount(contractEvents, address, latestContractSnapshot.Amount);
+
+            var contractSnapshot = new ContractSnapshot(
+                blockHeight,
+                address,
+                latestContractSnapshot.ContractName,
+                moduleReferenceEvent?.ModuleReference ?? latestContractSnapshot.ModuleReference,
+                amount,
+                source
+            );
+
+            await repository.AddAsync(contractSnapshot);
+        }
+    }
+
+    private static async Task AddInitialContractSnapshots(
+        IEnumerable<(ulong ContractAddressIndex, ulong ContractAddressSubIndex)> addressesInitialized,
+        IList<ContractEvent> contractEventsAdded,
+        IContractRepository repository, 
+        IModuleReadonlyRepository moduleReadonlyRepository,
+        ImportSource source
+        )
+    {
+        foreach (var contractAddress in addressesInitialized)
+        {
+            var address = new ContractAddress(contractAddress.ContractAddressIndex, contractAddress.ContractAddressSubIndex);
+            var contractEvents = contractEventsAdded
+                .Where(ce => ce.ContractAddressIndex == contractAddress.ContractAddressIndex && ce.ContractAddressSubIndex == contractAddress.ContractAddressSubIndex)
+                .OrderBy(ce => ce.BlockHeight)
+                .ThenBy(ce => ce.TransactionIndex)
+                .ThenBy(ce => ce.EventIndex)
+                .ToList();
+
+            var contractInitializationEvent = (contractEvents.First().Event as ContractInitialized)!;
+            var amount = GetAmount(contractEvents, address, 0);
+            
+            var blockHeight = contractEvents.Last().BlockHeight;
+            
+            var moduleReferenceEvent = GetLatestModuleReferenceContractLinkEvent(address, moduleReadonlyRepository)!;
+
+            var contractSnapshot = new ContractSnapshot(
+                blockHeight,
+                address,
+                contractInitializationEvent.GetName(),
+                moduleReferenceEvent.ModuleReference,
```

Comment:
The `contractInitializationEvent` should contain a module reference, not sure what `GetLatestModuleReferenceContractLinkEvent` does in this case

Reply:
The contractInitializationEvent only contains a link to the first module which was linked, but not necessarily the latest one. Hence we need to take the latest module link event.

Reply:
I am trying to figure out when `GetLatestModuleReferenceContractLinkEvent` is null, since we don't fall back to the initial module in this case as we do in the update case

Reply:
In the case where initialization has happened GetLatestModuleReferenceContractLinkEvent should never be null and above will give an exception. We would be failing fast and would be able to debug if this scenario happened before any data corruption occurred.